### PR TITLE
[GOBBLIN-746] Async loading FlowSpec

### PR DESCRIPTION
**Pull request by autumnust**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

This PR contains: 
- Change `SpecStore` interface to add listing-like methods, returning an iterator of all `Spec`s it contains. 
- Loading and scheduling `FlowSpec` in `GobblinServiceJobScheduler` asynchronously and notifying the completeness using a countdownlatch. 
- Add unit tests for `FSSpecStore` and `GobblinServiceJobScheduler`.

### JIRA
- [x] My PR addresses the following [Gobblin JIRA]
    - https://issues.apache.org/jira/browse/GOBBLIN-746

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by autumnust** [comment]
@sv2000 Can you help on review ? Thanks.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java [hunk 1]
```diff
@@ -136,6 +136,46 @@ public synchronized void setActive(boolean isActive) {
...
+        //Disable FLOW_RUN_IMMEDIATELY on service startup or leadership change
+        if (spec instanceof FlowSpec) {
+          Spec modifiedSpec = disableFlowRunImmediatelyOnStart((FlowSpec) spec);
+          onAddSpec(modifiedSpec);
+        } else {
+          onAddSpec(spec);
+        }
+      }
+    } finally {
+      flowSpecInitFinished.countDown();
```
Comment (sv2000) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java]: Is this countdown latch being used only in the test case? Your test case is waiting for at most 2 secs anyway. Can you simply do an AssertWithBackoff in your test case?

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/SpecSerDeException.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ */
+
+package org.apache.gobblin.runtime.api;
+
+import java.net.URI;
+
+/**
+ * An exception when {@link Spec} cannot be correctly serialized/deserialized from underlying storage.
+ */
+public class SpecSerDeException extends Exception{
```
[the remark this answers is not in the document]
Comment (autumnust) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/SpecSerDeException.java]: Fixed.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java (hunk 1 above)
Comment (autumnust) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/scheduler/GobblinServiceJobScheduler.java]: Correct it is mostly for testing purpose. I was thinking `CountDownLatch` might be the right way to test asynchronous service start. Thanks for point this out.

**Comment by autumnust** [comment]
@sv2000 Addressed comments and fix travis. Please take another look.

**Pull request closed without merging**